src/lib/provable.ts

```diff
@@ -0,0 +1,596 @@
+    return clone(type, condition.toBoolean() ? x : y);
+  }
+
+  // if b is variable, we compute if as follows:
+  // if(b, x, y)[i] = b*(x[i] - y[i]) + y[i]
+  let fields = xs.map((xi, i) => ifField(b, xi, ys[i]));
+  let aux = auxiliary(type, () => (condition.toBoolean() ? x : y));
+  return type.fromFields(fields, aux);
+}
+
+function ifImplicit<T extends ToFieldable>(condition: Bool, x: T, y: T): T {
+  let type = x.constructor;
+  if (type === undefined)
+    throw Error(
+      `You called Provable.if(bool, x, y) with an argument x that has no constructor, which is not supported.\n` +
+        `If x, y are Structs or other custom types, you can use the following:\n` +
+        `Provable.if(bool, MyType, x, y)`
+    );
+  if (type !== y.constructor)
+    throw Error(
+      'Provable.if: Mismatched argument types. Try using an explicit type argument:\n' +
+        `Provable.if(bool, MyType, x, y)`
+    );
+  if (!('fromFields' in type && 'toFields' in type)) {
+    throw Error(
+      'Provable.if: Invalid argument type. Try using an explicit type argument:\n' +
+        `Provable.if(bool, MyType, x, y)`
+    );
+  }
+  return ifExplicit(condition, type as any as Provable<T>, x, y);
+}
+
+function switch_<T, A extends FlexibleProvable<T>>(
+  mask: Bool[],
+  type: A,
+  values: T[]
+): T {
+  // picks the value at the index where mask is true
+  let nValues = values.length;
+  if (mask.length !== nValues)
+    throw Error(
+      `Provable.switch: \`values\` and \`mask\` have different lengths (${values.length} vs. ${mask.length}), which is not allowed.`
+    );
+  let checkMask = () => {
+    let nTrue = mask.filter((b) => b.toBoolean()).length;
+    if (nTrue > 1) {
+      throw Error(
+        `Provable.switch: \`mask\` must have 0 or 1 true element, found ${nTrue}.`
+      );
+    }
+  };
+  if (mask.every((b) => b.toField().isConstant())) checkMask();
+  else Provable.asProver(checkMask);
+  let size = type.sizeInFields();
+  let fields = Array(size).fill(Field(0));
+  for (let i = 0; i < nValues; i++) {
+    let valueFields = type.toFields(values[i]);
+    let maskField = mask[i].toField();
+    for (let j = 0; j < size; j++) {
+      let maybeField = valueFields[j].mul(maskField);
+      fields[j] = fields[j].add(maybeField);
+    }
+  }
+  let aux = auxiliary(type as Provable<T>, () => {
+    let i = mask.findIndex((b) => b.toBoolean());
+    if (i === -1) return undefined;
+    return values[i];
+  });
+  return (type as Provable<T>).fromFields(fields, aux);
+}
+
+// runners for provable code
+
+function log(...args: any) {
+  Provable.asProver(() => {
+    let prettyArgs = [];
+    for (let arg of args) {
+      if (arg?.toPretty !== undefined) prettyArgs.push(arg.toPretty());
+      else {
+        try {
+          prettyArgs.push(JSON.parse(JSON.stringify(arg)));
+        } catch {
+          prettyArgs.push(arg);
+        }
+      }
+    }
+    console.log(...prettyArgs);
+  });
+}
+
+function asProver(f: () => void) {
```

Comment:
everything below in this file is moved here without changes, except the new `checkLength` and `clone` helpers